File: ejercicios/fase-1/cliente-fetch-status/solucion.js
/**
 * Starter del ejercicio — Primero-Sin-IA.
 *
 * Implementa `obtenerNombreUsuario` a mano, sin IA. NO cambies las firmas ni los
 * nombres de las clases de error: los tests de `solucion.test.js` dependen de ellos.
 *
 * El corazón del ejercicio NO es "llamar fetch" (no se llama el fetch real): es
 *   1) validar el input ANTES de gastar una petición,
 *   2) separar el error de RED (la promesa rechaza) del status de error (la promesa
 *      RESUELVE con un status malo — fetch no rechaza por 404/500), y
 *   3) usar el DOBLE await (la respuesta, y luego .json()).
 *
 * `fetchFn` está INYECTADO: tus tests pasan una función falsa, sin tocar la red.
 *
 * Lee la sección 4.6 de la lección (1.7 JavaScript moderno) si te trabas.
 */

// Errores de dominio (ya definidos — úsalos donde el contrato lo pide).
export class EntradaInvalida extends Error {}
export class ServicioInalcanzable extends Error {}
export class UsuarioNoEncontrado extends Error {}
export class ServicioCaido extends Error {}
export class RespuestaInesperada extends Error {}

/**
 * Devuelve el nombre del usuario, manejando todos los modos de fallo.
 *
 * Contrato:
 *   - userId inválido (≤ 0 o no entero) → lanza EntradaInvalida ANTES de llamar fetchFn.
 *   - fetchFn RECHAZA (error de red simulado) → lanza ServicioInalcanzable.
 *   - status 200 → devuelve (await resp.json()).name
 *   - status 404 → lanza UsuarioNoEncontrado
 *   - status >= 500 → lanza ServicioCaido
 *   - cualquier otro status → lanza RespuestaInesperada
 *
 * @param {number} userId
 * @param {(userId:number) => Promise<{status:number, json:() => Promise<any>}>} fetchFn
 *        función inyectada que simula fetch. Devuelve una promesa de un objeto con
 *        `.status` (number) y `.json()` (async). Puede RECHAZAR (error de red).
 * @returns {Promise<string>}
 */
export async function obtenerNombreUsuario(userId, fetchFn) {
  throw new Error("Implementa esta función a mano, sin IA.");
}

File: ejercicios/fase-1/cliente-api-resiliente/solucion.py
"""Starter del ejercicio — Primero-Sin-IA.

Implementa `nombre_de_usuario` a mano, sin IA. NO cambies su firma: los tests de
`test_solucion.py` dependen de ella. Las excepciones de dominio ya están definidas.

La clave del ejercicio: `fetch` está INYECTADO. Tu función no sabe (ni le importa)
de dónde viene la respuesta — solo decide qué hacer con ella. Eso la hace testeable
sin tocar la red. Lee la sección 4.8 de la lección si te trabas.
"""

from __future__ import annotations

from typing import Any, Callable, Protocol


class Respuesta(Protocol):
    """Lo MÍNIMO que tu función necesita de una respuesta HTTP.

    Tanto una respuesta real de httpx como la respuesta falsa de los tests
    cumplen este contrato: tienen `.status_code` y un método `.json()`.
    """

    status_code: int

    def json(self) -> Any: ...


# `fetch` recibe un user_id y devuelve algo que cumple el Protocol Respuesta.
Fetch = Callable[[int], Respuesta]


class ServicioInalcanzable(Exception):
    """No hubo respuesta: error de red (timeout, conexión rechazada, sin internet)."""


class UsuarioNoEncontrado(Exception):
    """El servidor respondió 404: el usuario no existe."""


class ServicioCaido(Exception):
    """El servidor respondió 5xx: falló del lado del servidor."""


class RespuestaInesperada(Exception):
    """El servidor respondió con un status que no sabemos manejar."""


def nombre_de_usuario(user_id: int, fetch: Fetch) -> str:
    """Devuelve el nombre del usuario, manejando cada modo de fallo.

    Ver el contrato completo en el README.md de este ejercicio.
    """
    raise NotImplementedError("Implementa esta función a mano, sin IA.")


if __name__ == "__main__":
    # Demostración con un `fetch` falso (sin red). Predice qué imprime antes de correr.
    class _RespFalsa:
        def __init__(self, status_code: int, payload: Any) -> None:
            self.status_code = status_code
            self._payload = payload

        def json(self) -> Any:
            return self._payload

    def fetch_demo(user_id: int) -> _RespFalsa:
        return _RespFalsa(200, {"name": "Ada Lovelace"})

    print(nombre_de_usuario(1, fetch_demo))
